Firestore/Source/Model/FSTFieldValue.h

```diff
@@ -91,7 +124,7 @@ typedef NS_ENUM(NSInteger, FSTTypeOrder) {
 + (instancetype)trueValue;
 + (instancetype)falseValue;
 + (instancetype)booleanValue:(BOOL)value;
-- (NSNumber *)value;
+- (NSNumber *)valueWithOptions:(FSTFieldValueOptions *)options;
```

Comment:
So the old versions of these classes included a covariant type of the return type of `value`.

Were we making use of that? I worry that this still builds because `id` as a return type auto-casts itself to anything you assign it to.

Do we need the covariant type? If so we probably should continue to declare `value` with covariant types. If not we probably don't need to redeclare these.

In any case, the inconsistency between `value` and `valueWithOptions:` is jarring.

Reply:
I'm fine with adding these back to the header files. The reason I removed them is that they spew a bunch of warnings about unimplemented methods (which can be suppressed). Let me know what you lean towards.

Reply:
The thing I was asking for here was how we're using the result of `value`. For example, are we actually using the fact that `-[FSTIntegerValue value]` returns `NSNumber` anywhere?

Reply:
We do care here: https://github.com/Firebase/firebase-ios-sdk/blob/4fbc96904fe86b3a0cdb91bfa6d96491221eb6b9/Firestore/Source/Core/FSTQuery.m#L126

and here: https://github.com/firebase/firebase-ios-sdk/blob/4fbc96904fe86b3a0cdb91bfa6d96491221eb6b9/Firestore/Source/Remote/FSTSerializerBeta.m#L163

Reply:
This change makes all those call sites effectively untyped because they're all now using `id` as the return type for `-value`. This only compiles because you can cast freely to and from `id`.

The core concern raised in this comment remains: `-value` and `-valueWithOptions:` should have the same return type.

Probably this means FSTFieldValue should have a lightweight generic type so that the return type of `-value` can be generic even if the implementation actually just returns `id`.

Reply:
I changed this to use Objective-C Generics.

I tried this earlier and abandoned the approach since App Code shows a warning ("Casting to probable incompatible type") when casting from FSTFieldValue to any of the implementing classes. This doesn't show up in the actual compiler warnings though and seems to be an AppCode-only issue.